### ISSUE TITLE
no-fn-reference-in-iterator: Ignore locally declared functions?

### ISSUE THREAD

**Issue by coreyfarrell** [sindresorhus/eslint-plugin-unicorn#717]
I would like the following to be ignored by `no-fn-reference-in-iterator`:
```js
const fn = item => {
  // do stuff with iterator items
};

iter1.forEach(fn);
iter2.forEach(fn);
iter3.forEach(fn);
```

Since `fn` is locally declared and defined the purpose of `no-fn-reference-in-iterator` does not apply.  I see value in the following situations being reported:
```js
const mapper = require('unicorn-mapper');

function doStuff(iter, fn) {
  iter.forEach(fn); // error on `fn` provided by argument
}

[2, 3, 4].forEach(mapper); // error as `mapper` is from a `require`
```

**Comment by sindresorhus** [comment]
I think the existing behavior is correct. Even if the `fn` function is in the same scope, it could be far from the `forEach`. It's also about readability.

We could maybe add an option to ignore functions in the same scope.

**Comment by sindresorhus** [comment]
One case we should ignore is `.map(destination ? toFile : toBuffer)`, since then it's clear that it was the intended callback function, just that it had to be conditional.

**Comment by coreyfarrell** [comment]
I would certainly appreciate an option to ignore functions that are declared in the local scope (or above).  In real life I would use a more descriptive name than `fn`, so I would use naming to make things clear rather than doing `.map(item => fn(item))`.  A lot of my code targets the browser and `superLongAndDescriptiveFunctionName` gets reduced to one or two characters by terser but the technically unneeded arrow function inside `.map()` cannot be automatically stripped.  I know it's a tiny difference but many small sacrifices of size add up in a large web project.  Not trying to convince you that I'm right just sharing where I'm approaching this from.

With regard to ignoring `.map(destination ? toFile : toBuffer)` would that still warn if one or both of those functions are not locally declared?

**Comment by sindresorhus** [comment]
> With regard to ignoring .map(destination ? toFile : toBuffer) would that still warn if one or both of those functions are not locally declared?

Yes

**Comment by fregante** [comment]
I've been disabling this rule as I go.

I get that it tries to prevent future breakage, but this exclusively depends on the function signature:

- does `fn` accept multiple arguments? Report it.

That's it.

Only `node_modules` introduce uncertainty because a future `npm install` might change the signature of `fn` after the lint, unless… 

- there's a lockfile which guarantees that `fn` won't just change under your feet.

---

This means that single-parameter local functions (same file or other file) should **not** be reported by this check. If a local `fn` is changed a multi-parameter function, the lint will likely be run after this change and report it.

If lockfile detection can be part of a rule, `node_modules` can also follow the same logic.

**Comment by coreyfarrell** [comment]
> I get that it tries to prevent future breakage, but this exclusively depends on the function signature:
> 
> * does `fn` accept multiple arguments? Report it.

Not always true that multiple arguments is a problem:
```js
const arrayItemDump = (value, index) => console.log(`item[${index}]: ${value}`);

list.forEach(arrayItemDump);
```

This is why I proposed an option to ignore locally declared functions but did not suggest that it should matter how many arguments the function takes.

> * the function signature cannot be verified? Report it.

Note that eslint rules cannot (reasonably) know the signature of functions exported by any other file, ESLint processes on a per-file basis so it's very difficult and error prone for a rule to make decisions based on contents of other files.  Cross-file analysis tends to break any eslint result caching as well.

> Only `node_modules` introduce uncertainty because a future `npm install` might change the signature of `fn` after the lint, unless…
> 
> * there's a lockfile which guarantees that `fn` won't just change under your feet.
> 
> 
> This means that **single-parameter local functions** (same file or other file) should **not** be reported by this check. If a local `fn` is changed to a **multi-parameter**, the lint will likely be run after this change and **report it**.
> 
> If lockfile detection can be part of a rule, `node_modules` can also follow the same logic.

package-lock.json is only for your local development install.  Users who install your package do not get your lockfile so in this context the lockfile would only prevent your tests from catching an error caused by an updated dependency.

**Comment by fregante** [comment]
> package-lock.json is only for your local development install.

Let's say that lockfiles are only for apps so, yeah, they can't be used as an exception.

> Note that eslint rules cannot (reasonably) know the signature of functions exported by any other file

I don't understand this, `eslint-plugin-import` even reports dependencies not defined in package.json, so this should be possible… except maybe that it's relatively slow.

Also perhaps function types are available in TypeScript projects.

> Not always true that multiple arguments is a problem:

You're right. Then I think this `rule` just gets in the way of functional programming and I find it hard to justify it, IMHO.

**Comment by fregante** [comment]
> > Not always true that multiple arguments is a problem:
> 
> You're right. Then I think this `rule` just gets in the way of functional programming and I find it hard to justify it, IMHO.

Changed my mind on this specific point:

```diff
- const showFruits = (name) => console.log(`I have many ${name}`)
+ const showFruits = (name, count) => console.log(`I have exactly ${count} ${name}`)

list.forEach(showFruits);
```

The problem is that (regardless of types) the linter does not know what the expected input of the iterator is. The second parameter could match the exact type and pass everywhere, yet still fall into the trap that this rule is meant to prevent.

**Comment by sindresorhus** [comment]
I think the current behavior should stay as is in `no-array-callback-reference`.

The rule's guarantee is intentionally simple: callback references can accidentally receive `index` and `array`, and that risk also exists for locally declared functions. A local one parameter callback can later grow another parameter or be reused somewhere else, so exempting locals would weaken the exact behavior the rule is meant to protect.

I do not think an `ignoreLocalFunctions` option or arity/signature analysis is worth it here. It adds complexity, is still incomplete without type and cross file analysis, and makes the rule harder to explain. The practical fix is to use an inline wrapper when you want to lock the passed arguments:

```js
array.map(value => callback(value));
```

For users who prefer direct callback references, the rule is currently `unopinionated`, so opting out through config is the intended escape hatch.
